python/ray/data/_internal/lazy_block_list.py

```diff
@@ -566,25 +613,58 @@ def _num_computed(self) -> int:
         return i
 
 
-def _execute_read_task(
+def _execute_read_task_nosplit(
     i: int,
     task: ReadTask,
     context: DatasetContext,
     stats_uuid: str,
     stats_actor: ray.actor.ActorHandle,
-) -> Tuple[MaybeBlockPartition, BlockPartitionMetadata]:
+) -> Tuple[Block, BlockMetadata]:
     DatasetContext._set_current(context)
     stats = BlockExecStats.builder()
 
     # Execute the task.
-    block = task()
+    block = next(task())
 
     metadata = task.get_metadata()
-    if context.block_splitting_enabled:
-        metadata.exec_stats = stats.build()
-    else:
-        metadata = BlockAccessor.for_block(block).get_metadata(
-            input_files=metadata.input_files, exec_stats=stats.build()
-        )
+    metadata = BlockAccessor.for_block(block).get_metadata(
+        input_files=metadata.input_files, exec_stats=stats.build()
+    )
     stats_actor.record_task.remote(stats_uuid, i, metadata)
     return block, metadata
+
+
+def _execute_read_task_split(
+    i: int,
+    task: ReadTask,
+    context: DatasetContext,
+    stats_uuid: str,
+    stats_actor: ray.actor.ActorHandle,
+) -> Iterable[Union[Block, List[BlockMetadata]]]:
+    """Execute read task with dynamic block splitting.
+
+    Returns an Iterable of blocks followed by their metadata.
+    Example of return value for 3 blocks:
+    (Block1, Block2, Block3, [BlockMetadata1, BlockMetadata2, BlockMetadata3])
+    """
+    DatasetContext._set_current(context)
+
+    # Execute the task.
+    blocks = task()
+
+    task_metadata = task.get_metadata()
+    task_metadata.exec_stats = BlockExecStats()
+    blocks_metadata = []
+    block_exec_stats = BlockExecStats.builder()
+    for block in blocks:
+        metadata = BlockAccessor.for_block(block).get_metadata(
+            input_files=task_metadata.input_files, exec_stats=block_exec_stats.build()
+        )
+        yield block
+        blocks_metadata.append(metadata)
+        task_metadata.exec_stats.add(metadata.exec_stats)
```

Comment:
Do we want to do the same for the integration in the `TaskPoolStrategy`, in `compute.py`?

Reply:
This task_metadata is the same as blockmetadata, so seems stats actor is indeed bookkeeping some duplicate info?

Reply:
> This task_metadata is the same as blockmetadata, so seems stats actor is indeed bookkeeping some duplicate info?

given task metadata has aggregated execution stats from each block, why this task metadata is the same as block metadata?

Reply:
The stats kept in the actor can be derived from the block metadata, so if block metadata are kept around the task metadata has no need to be kept?

Reply:
It seems like we have a few options for how to report and ultimately present read stats:
1. **(Current)** We can aggregate execution stats from all blocks in this task into a single "task execution stats", that we then report to the stats actor once done reading all blocks.
2. We can send the per-block execution stats to the stats actor, either as the block is read (move `stats_actor.record_task.remote()` into the for loop) or at the end as a list, and then the stats actor can either aggregate these into "task execution stats" or store these as individual block stats. The latter would require some stats actor refactoring to hold a list of block execution stats per read task, but would then give us per-block execution stats granularity.

In terms of stats reporting, I think that (2) with the per-block execution stats stored and reported as per-block execution stats would better match what [`TaskPoolStrategy` is doing](https://github.com/ray-project/ray/blob/0f705888c6fa7106e78b777aff07c72aed639c62/python/ray/data/_internal/compute.py#L411-L423) for e.g. `.map_batches()` tasks, and I think is what we want to do since our stats are block-centric (i.e. "stats for producing this block") rather than task-centric (i.e. "stats for executing this task").

What do y'all think of going with that version of (2), with all block stats reported at the end of the read task, to keep things consistent? We can reexamine this with the upcoming AIR observability overhaul, cc @gjoliver

Reply:
@clarkzinzow - I think (2).report per-block stats to stats actor makes sense to me, and should be better (per-block & align with `map_batches` stage). Let me make the change accordingly.

Reply:
@jianoaix - `_StatsActor` has only one API to record stats - `record_task()`. There's no duplicated info being recorded there. So the discussion here is whether (1). recording one single metadata per task, aggregating all blocks metadata belonging to the task, or (2). recording the blocks metadata separately as a list. Here (2) makes more sense to me.

Reply:
@clarkzinzow, @jianoaix - updated the code to report per-block stats to stats actor.